modules/xfeatures2d/src/beblid.cpp

```diff
@@ -0,0 +1,602 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+// Author: Iago Suarez <iago.suarez.canosa@alumnos.upm.es>
+
+// Implementation of the article:
+//     Iago Suarez, Ghesn Sfeir, Jose M. Buenaposada, and Luis Baumela.
+//     BEBLID: Boosted Efficient Binary Local Image Descriptor.
+//     Pattern Recognition Letters, 133:366–372, 2020.
+
+#include <bitset>
+#include "precomp.hpp"
+
+#define CV_BEBLID_PARALLEL
+
+#define CV_ROUNDNUM(x) ((int)(x + 0.5f))
+#define CV_DEGREES_TO_RADS 0.017453292519943295 // (M_PI / 180.0)
+#define CV_BEBLID_EXTRA_RATIO_MARGIN 1.75f
+
+using namespace cv;
+using namespace std;
+
+
+namespace cv
+{
+namespace xfeatures2d
+{
+
+// Struct containing the 6 parameters that define an Average Box weak-learner
+struct ABWLParams
+{
+    int x1, y1, x2, y2, boxRadius, th;
+};
+
+// BEBLID implementation
+class BEBLID_Impl CV_FINAL: public BEBLID
+{
+public:
+
+    // constructor
+    explicit BEBLID_Impl(float scale_factor, int n_bits = SIZE_512_BITS);
+
+    // destructor
+    ~BEBLID_Impl() CV_OVERRIDE = default;
+
+    // returns the descriptor length in bytes
+    int descriptorSize() const CV_OVERRIDE { return int(wl_params_.size() / 8); }
+
+    // returns the descriptor type
+    int descriptorType() const CV_OVERRIDE { return CV_8UC1; }
+
+    // returns the default norm type
+    int defaultNorm() const CV_OVERRIDE { return cv::NORM_HAMMING;  }
+
+    // compute descriptors given keypoints
+    void compute(InputArray image, vector<KeyPoint> &keypoints, OutputArray descriptors) CV_OVERRIDE;
+
+private:
+    std::vector<ABWLParams> wl_params_;
+    float scale_factor_;
+    cv::Size patch_size_;
+
+    void computeBEBLID(const cv::Mat &integralImg,
+                       const std::vector<cv::KeyPoint> &keypoints,
+                       cv::Mat &descriptors);
+}; // END BEBLID_Impl CLASS
+
+/**
+ * @brief Function that determines if a keypoint is close to the image border.
+ * @param kp The detected keypoint
+ * @param imgSize The size of the image
+ * @param patchSize The size of the normalized patch where the measurement functions were learnt.
+ * @param scaleFactor A scale factor that magnifies the measurement functions w.r.t. the keypoint.
+ * @return true if the keypoint is in the border, false otherwise
+ */
+inline bool isKeypointInTheBorder(const cv::KeyPoint &kp,
```

Comment:
Please use `static` / `static inline` for local functions (here and below)

Reply:
done